pkg/clustermesh/mcsapi/serviceexportsync_controller.go

```diff
@@ -0,0 +1,199 @@
+)
+
+// ServiceExportResource builds the Resource[ServiceExport] object.
+func ServiceExportResource(lc cell.Lifecycle, cs client.Clientset, opts ...func(*metav1.ListOptions)) resource.Resource[*mcsapiv1alpha1.ServiceExport] {
+	if !cs.IsEnabled() {
+		return nil
+	}
+	lw := utils.ListerWatcherWithModifiers(
+		utils.ListerWatcherFromTyped(cs.MulticlusterV1alpha1().ServiceExports("")),
+		opts...,
+	)
+	return resource.New[*mcsapiv1alpha1.ServiceExport](lc, lw, resource.WithMetric("ServiceExport"))
+}
+
+type mcsAPIServiceSyncSpecSyncController struct {
+	logger logrus.FieldLogger
+
+	clusterName  string
+	kvs          store.SyncStore
+	kvstoreReady chan struct{}
+	queue        workqueue.RateLimitingInterface
+
+	serviceStore       resource.Store[*slim_corev1.Service]
+	serviceExportStore resource.Store[*mcsapiv1alpha1.ServiceExport]
+}
+
+func newMCSAPIServiceSyncSpecSyncController(logger logrus.FieldLogger, clusterName string) *mcsAPIServiceSyncSpecSyncController {
+	return &mcsAPIServiceSyncSpecSyncController{
+		logger:       logger,
+		clusterName:  clusterName,
+		kvstoreReady: make(chan struct{}),
+		queue:        workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "mcsapispec-sync-queue"),
+	}
+}
+
+func (c *mcsAPIServiceSyncSpecSyncController) Run(ctx context.Context, serviceExportResource resource.Resource[*mcsapiv1alpha1.ServiceExport], wg *sync.WaitGroup, cfg watchers.ServiceSyncParameters) {
+	defer c.queue.ShutDown()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		if cfg.Backend == nil {
+			// Needs to be assigned in a separate goroutine, since it might block
+			// if the client is not yet initialized.
+			cfg.Backend = kvstore.Client()
+		}
+
+		c.kvs = cfg.StoreFactory.NewSyncStore(cfg.ClusterInfo.Name,
+			cfg.Backend, ServiceExportStorePrefix)
+		close(c.kvstoreReady)
+		c.kvs.Run(ctx)
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		serviceEvents := cfg.Services.Events(ctx)
+		serviceStore, err := cfg.Services.Store(ctx)
+		c.serviceStore = serviceStore
+		if err != nil {
+			c.logger.Errorf("can't init service store: %s", err)
+			return
+		}
+		serviceExportsEvents := serviceExportResource.Events(ctx)
+		c.serviceExportStore, err = serviceExportResource.Store(ctx)
+		if err != nil {
+			c.logger.Errorf("can't init service export store: %s", err)
+			return
+		}
+
+		servicesSynced, serviceExportsSynced := false, false
+
+		for serviceEvents != nil || serviceExportsEvents != nil {
+			select {
+			case ev, ok := <-serviceEvents:
+				if !ok {
+					serviceEvents = nil
+					continue
+				}
+
+				if ev.Kind == resource.Sync {
+					servicesSynced = true
+					if servicesSynced && serviceExportsSynced {
+						c.startProcessing(ctx, wg, cfg)
+					}
+					ev.Done(nil)
+					continue
+				}
+				if !servicesSynced || !serviceExportsSynced {
+					continue
+				}
+
+				c.queue.Add(resource.Key{
+					Namespace: ev.Key.Namespace,
+					Name:      ev.Key.Name,
+				})
```

Comment:
Ohh I didn't catch that getting a store was waiting for a full sync, then yes I suppose that we could do without a queue then :eyes:! The only thing is that with a queue I could de-duplicate updates but considering that ServiceExport has nothing past name and namespace it would probably only be somewhat useful when the service starts :thinking:.

Reply:
De-duplication is also already performed by the working queue of sync store, hence it would not add any additional benefit.